### PR TITLE
Adding Hello Reshape Python sample

### PR DESCRIPTION
**Pull request by mmakridi**

### CODE REVIEW

File: inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    # --------------------------- 4. Configure input & output ---------------------------------------------
+    # --------------------------- Prepare input blobs -----------------------------------------------------
+    log.info("Preparing input blobs")
+    assert (len(net.input_info.keys()) == 1 or len(
+        net.input_info.keys()) == 2), "Sample supports topologies only with 1 or 2 inputs"
+    out_blob = next(iter(net.outputs))
+    input_name, input_info_name = "", ""
+
+    for input_key in net.input_info:
+        if len(net.input_info[input_key].layout) == 4:
```
Comment (mryzhov) [inline inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py:117]: Can you please comment on such magic digits at least or use understandable variables or constants.
for example, tensor_dementions = 4

**Review by mryzhov — CHANGES_REQUESTED** [review]

File: inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+            if proposal[2] > 0.5:
+                print(" WILL BE PRINTED!")
+                if not imid in boxes.keys():
+                    boxes[imid] = []
+                boxes[imid].append([xmin, ymin, xmax, ymax])
+                if not imid in classes.keys():
+                    classes[imid] = []
+                classes[imid].append(label)
+            else:
+                print()
```
Comment (generalova-kate) [inline inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py:194]: why do you need such strange `else`? I think you can add an empty row to any place without if-else construction.

File: inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+            label = np.int(proposal[1])
+            confidence = proposal[2]
+            xmin = np.int(iw * proposal[3])
+            ymin = np.int(ih * proposal[4])
+            xmax = np.int(iw * proposal[5])
+            ymax = np.int(ih * proposal[6])
+            print("[{},{}] element, prob = {:.6}    ({},{})-({},{}) batch id : {}" \
+                  .format(number, label, confidence, xmin, ymin, xmax, ymax, imid), end="")
+            if proposal[2] > 0.5:
+                print(" WILL BE PRINTED!")
```
Comment (generalova-kate) [inline inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py:186]: is it debug print? please remove or update text without capslock.

File: inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+                           "Sample will look for a suitable plugin for device specified (CPU by default)",
+                      default="CPU", type=str)
+    args.add_argument("--labels", help="Optional. Labels mapping file", default=None, type=str)
+    args.add_argument("-nt", "--number_top", help="Optional. Number of top results", default=10, type=int)
+
+    return parser
+
+
+def main():
+    log.basicConfig(format="[ %(levelname)s ] %(message)s", level=log.INFO, stream=sys.stdout)
```
Comment (generalova-kate) [inline inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py:52]: I think we should replace all prints with logger messages if you use it.

File: inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+    res = exec_net.infer(inputs=data)
+    # -----------------------------------------------------------------------------------------------------
+
+    # --------------------------- Read and postprocess output ---------------------------------------------
+    log.info("Processing output blobs")
+    res = res[out_blob]
+    boxes, classes = {}, {}
+    data = res[0][0]
+    for number, proposal in enumerate(data):
+        if proposal[2] > 0:
```
Comment (generalova-kate) [inline inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py:174]: proposal[2], res = res[out_blob], data = res[0][0] - please improve naming

File: inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+    output_dims = output_info.shape
+    if len(output_dims) != 4:
+        log.error("Incorrect output dimensions for SSD model")
+    max_proposal_count, object_size = output_dims[2], output_dims[3]
+
+    if object_size != 7:
+        log.error("Output item should have 7 as a last dimension")
+
+    output_info.precision = "FP32"
```
Comment (generalova-kate) [inline inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py:158]: Please add info to README or some other place, that output will be FP32

File: inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+def build_argparser():
+    parser = ArgumentParser(add_help=False)
+    args = parser.add_argument_group("Options")
+    args.add_argument('-h', '--help', action='help', default=SUPPRESS, help='Show this help message and exit.')
+    args.add_argument("-m", "--model", help="Required. Path to an .xml or .onnx file with a trained model.",
+                      required=True, type=str)
+    args.add_argument("-i", "--input", help="Required. Path to an image file.",
+                      required=True, type=str)
+    args.add_argument("-l", "--cpu_extension",
```
[the remark this answers is not in the document]
Comment (mmakridi) [reply inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py]: Removed, thanks

File: inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+                      help="Optional. Required for CPU custom layers. "
+                           "Absolute path to a shared library with the kernels implementations.",
+                      type=str, default=None)
+    args.add_argument("-d", "--device",
+                      help="Optional. Specify the target device to infer on; "
+                           "CPU, GPU, FPGA or MYRIAD is acceptable. "
+                           "Sample will look for a suitable plugin for device specified (CPU by default)",
+                      default="CPU", type=str)
+    args.add_argument("--labels", help="Optional. Labels mapping file", default=None, type=str)
+    args.add_argument("-nt", "--number_top", help="Optional. Number of top results", default=10, type=int)
```
[the remark this answers is not in the document]
Comment (mmakridi) [reply inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py]: Removed, thanks

File: inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+    print("inputs number: " + str(len(net.input_info.keys())))
+    assert len(net.input_info.keys()) == 1, 'Sample supports clean SSD network with one input and one output'
+    input_name = list(net.input_info.keys())[0]
+
+    for input_key in net.input_info:
+        print("input shape: " + str(net.input_info[input_key].input_data.shape))
+        print("input key: " + input_key)
+        if len(net.input_info[input_key].input_data.layout) == 4:
+            n, c, h, w = net.input_info[input_key].input_data.shape
+            assert n == 1, 'Sample only supports topologies with one input image'
```
[the remark this answers is not in the document]
Comment (mmakridi) [reply inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py]: Added, thanks

File: inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+
+    if args.cpu_extension and "CPU" in args.device:
+        ie.add_extension(args.cpu_extension, "CPU")
+        log.info("CPU extension loaded: {}".format(args.cpu_extension))
+    # -----------------------------------------------------------------------------------------------------
+
+    # --------------------------- 3. Read and preprocess input --------------------------------------------
+
+    print("inputs number: " + str(len(net.input_info.keys())))
+    assert len(net.input_info.keys()) == 1, 'Sample supports clean SSD network with one input and one output'
```
[the remark this answers is not in the document]
Comment (mmakridi) [reply inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py]: Added, thanks

File: inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+    args.add_argument("-l", "--cpu_extension",
+                      help="Optional. Required for CPU custom layers. "
+                           "Absolute path to a shared library with the kernels implementations.",
+                      type=str, default=None)
+    args.add_argument("-d", "--device",
+                      help="Optional. Specify the target device to infer on; "
+                           "CPU, GPU, FPGA or MYRIAD is acceptable. "
+                           "Sample will look for a suitable plugin for device specified (CPU by default)",
+                      default="CPU", type=str)
+    args.add_argument("--labels", help="Optional. Labels mapping file", default=None, type=str)
```
[the remark this answers is not in the document]
Comment (mmakridi) [reply inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py]: Removed, thanks

File: inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+    # --------------------------- 3. Read and preprocess input --------------------------------------------
+
+    print("inputs number: " + str(len(net.input_info.keys())))
+    assert len(net.input_info.keys()) == 1, 'Sample supports clean SSD network with one input and one output'
+    input_name = list(net.input_info.keys())[0]
+
+    for input_key in net.input_info:
+        print("input shape: " + str(net.input_info[input_key].input_data.shape))
+        print("input key: " + input_key)
+        if len(net.input_info[input_key].input_data.layout) == 4:
```
[the remark this answers is not in the document]
Comment (mmakridi) [reply inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py]: Added exception for != 4 case, thanks

File: inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+            n, c, h, w = net.input_info[input_key].input_data.shape
+            assert n == 1, 'Sample only supports topologies with one input image'
+    
+    h_new, w_new = cv2.imread(args.input).shape[:-1]
+    images = np.ndarray(shape=(n, c, h_new, w_new))
+    images_hw = []
+    for i in range(n):
```
[the remark this answers is not in the document]
Comment (mmakridi) [reply inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py]: Fixed, thanks

File: inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+    net.reshape({input_name: [n, c, h_new, w_new]})
+    print(net.input_info['data'].input_data.shape)
+
+    # -----------------------------------------------------------------------------------------------------
+
+    # --------------------------- 4. Configure input & output ---------------------------------------------
+    # --------------------------- Prepare input blobs -----------------------------------------------------
+    log.info("Preparing input blobs")
+    assert (len(net.input_info.keys()) == 1 or len(
+        net.input_info.keys()) == 2), "Sample supports topologies only with 1 or 2 inputs"
```
[the remark this answers is not in the document]
Comment (mmakridi) [reply inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py]: You are right, removed it here, thanks

File: inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py (hunk 1 above)
Comment (mmakridi) [reply inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py]: Added variable `supported_input_dims` and comment for it

File: inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py (hunk 2 above)
Comment (mmakridi) [reply inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py]: Refactored output, fixed

File: inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py (hunk 3 above)
Comment (mmakridi) [reply inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py]: Refactored output, fixed

File: inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py (hunk 4 above)
Comment (mmakridi) [reply inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py]: Switched to `log.info` everywhere, thanks

File: inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py (hunk 5 above)
Comment (mmakridi) [reply inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py]: Added comments clarifying some naming and fixed some other naming

File: inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py (hunk 6 above)
Comment (mmakridi) [reply inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py]: Looks like no samples readmes ever mention input/output precision, so I suppose we should not add it here for consistency. Moreover, samples are not designed for any performance/accuracy testing, so we do not care much about precision in samples anyway.

**Review by mryzhov — APPROVED** [review]
Looks good

File: inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+                if not imid in classes.keys():
+                    classes[imid] = []
+                classes[imid].append(label)
+            else:
+                print()
+
+    for imid in classes:
+        tmp_image = cv2.imread(args.input)
+        for box in boxes[imid]:
+            cv2.rectangle(tmp_image, (box[0], box[1]), (box[2], box[3]), (232, 35, 244), 2)
```
[the remark this answers is not in the document]
Comment (mmakridi) [reply inference-engine/ie_bridges/python/sample/hello_reshape_ssd/hello_reshape_ssd.py]: Fixed, thanks

**Pull request merged**